### DISCUSSION TITLE
Calculation of logistic functions gives not-a-number (nan) errors or zero when it should increase from zero to unity.

### DISCUSSION THREAD

**Discussion by EdSheltonUKAEA** [idaholab/moose#28289] [Q&A General]
### Check these boxes if you have followed the posting rules.

- [X] Q&A General is the most appropriate section for my question
- [X] I have consulted the posting Guidelines on the Discussions front page
- [X] I have searched the Discussions forum and my question has not been asked before
- [X] I have searched the MOOSE website and the documentation does not answer my question
- [X] I have formatted my post following the posting guidelines (screenshots as a last resort, triple back quotes around pasted text)

### Question

I am trying to use logistic functions to add a function as a smooth sigmoid over a time period.  But the calculation of the function behaves erratically.

The function I would like to use in a ParsedFunction is:
`expression = 'if(t<79500, 0, 1/(1 + exp((80000-t)/50) ) )'`

But even if I start my solve at t = 79500 s, it fails immediately with:
`Time Step 1, time = 79510, dt = 10
Pre-SMO residual: -nan

Performing automatic scaling calculation

Nonlinear solve did not converge due to DIVERGED_FNORM_NAN iterations 0
 Solve Did NOT Converge!`

At t = 79500, exp((80000-t)/50) = exp(10) = 22026 which shouldn't be causing the NAN error.

So I've tried an alternative expression for the logistic function (without an if statement since for t<<80000, the exp term should tend to zero):
`expression = '1 - (1/(1 + exp((t - 80000)/50)) )'`

This code runs, but the value of the function calculated with this expression stays stuck at zero, even when t = 80500 and higher, where it should tend to unity.  This makes even less sense to me than the nan error.

However, if I let my solve start at t = 0 and use:
`expression = '1/(1 + exp((500-t)/50) )'`
then everything works fine.  The functions increases sigmoidally up to unity at about t = 1380 s and beyond.

So I can "force" it to work in this way and then append this bit of the solve to previous output up to 79500 s, but this is rather inelegant and time-consuming.

Why can the function expression cope when the logistic function mid-point is at 500s, but not at 80000 s (for t>79500 s), even though the range of values of the argument of the exp function is the same?

Is there some setting I can change to get the solve to work around 80000 s?

**Comment by GiudGiud** [comment]
Hello

Thanks for reporting this. This must be an issue in the parser, we use this
http://warp.povusers.org/FunctionParser/fparser.html

Would you mind checking if they mention any overflow issue in their documentation?

Guillaume

**Reply by EdSheltonUKAEA** [reply]
Hi
Thanks for the quick reply.
I'm not sure what I'd be looking for in the document at that link as I'm not fluent in C++ or understanding all the details of the calculations that go on behind the scenes.
I sort of assume that MOOSE would "do the math" for the functions I put in.  Hence my confusion that it didn't.
Ed

**Reply by EdSheltonUKAEA** [reply]
It does say this, which seems it could be relevant:

> "(Note that currently the exponentiation operator is not supported for FunctionParser_li nor FunctionParser_gmpint. With the former the result would very easily overflow, making its usefulness questionable. With the latter it could be easily abused to make the program run out of memory; think of a function like "10^10^10^100000".) "

**Reply by EdSheltonUKAEA** [reply]
Mind you, that means exponentiation as raising to powers, not the exponential function.  So probably not relevant after all.

**Reply by EdSheltonUKAEA** [reply]
Here are the only other mentions of "overflow":

> However, the library can not guarantee that it will catch all possible floating point errors before performing them, because this is impossible to do with standard C++. For example, dividing a very large value by a value which is very close to zero, or calculating the logarithm of a very small value may overflow the result, as well as multiplying two very large values. Raising a negative number to a non-integral power may cause a NaN result, etc.
> 
> As a rule of thumb, the library will (by default) detect invalid operations if they are invalid for a range of values. For example, square root is undefined for all negative values, and arc sine is undefined only values outside the range [-1, 1]. In general, operations which are invalid for only one single value (rather than a contiguous range of values) will not be detected (division by the exact value of zero is an exception to this rule) nor will overflow/underflow situations.

**Reply by GiudGiud** [reply]
ok thanks for searching the documentation. There is not a strong hint that the exponentials are not evaluated as we would expect.

could you try this by any chance?
```
expression = '1 - (1/(1 + exp( ((t - 80000)/50)))  )'
``` 

In case it's optimizing by doing something like
exp(t/50) exp(- 80000)/50)

which would obviously be problematic

**Reply by EdSheltonUKAEA** [reply]
I'll aim to give this a try in the next day or two.  If I understand it right, including the extra set of brackets may force the parser to process the calculations in a different order.
But for now I'm able to carry on with my workaround, see below.

**Reply by GiudGiud** [reply]
That is what I am thinking yes
ok I am glad. We'll need to work in FParser to fix this. @dschwen thoughts?

**Reply by EdSheltonUKAEA** [reply]
Trying
`expression = '1 - (1/(1 + exp( ((t - 80000)/50)))  )'`
didn't work.
The value of the function stays at zero even for t = 84700, the end time of my model.

**Comment by EdSheltonUKAEA** [comment]
I'm going to try a workaround where I set time t in the MOOSE model to be the actual value starting at 79500, but define a function: 
tdash = t - 79500
which I can use in the argument of the exponential term of the logistic function.

But a more robust solution would be welcome if anyone can suggest one.

**Reply by EdSheltonUKAEA** [reply]
A quick update, this seems to have worked.  I've now managed to run my model to the end and the increasing sigmoidal logistic function worked perfectly.

**Reply by EdSheltonUKAEA** [reply]
For clarity, here's what I'm using which gives the desired output:
```
  [tdash_fn]
    type = ParsedFunction
    expression = 't-79500'  
  []
  [H_time_smooth_fn]
    type = ParsedFunction
    symbol_names = 'tdash'
    symbol_values = 'tdash_fn'
    expression = 'if(t<79500, 0, 1 - 1/(1+exp((tdash-500)/50)) )'
  []
```

**Reply by EdSheltonUKAEA** [reply]
Or a bit more straightforward:
```
  [tdash_fn]
    type = ParsedFunction
    expression = 't-80000'  
  []
  [H_time_smooth_fn]
    type = ParsedFunction
    symbol_names = 'tdash'
    symbol_values = 'tdash_fn'
    expression = 'if(t<79500, 0, 1 - 1/(1+exp(tdash/50)) )'
  []
```
